### PR TITLE
[WIP] Implements ELO rating for sunspotter

### LINKED ISSUE #7: Recreate the ELO rating.
The complexity score for the ARs may be compromised because of a mixup of datasets.
This needs to be verified before any training can be done.
We need to recreate the ELO rating and compare it with other ratings (e.g., [Glicko's](https://en.wikipedia.org/wiki/Glicko_rating_system),
  [Bradley-Terry's](https://en.wikipedia.org/wiki/Bradley%E2%80%93Terry_model)).

### PR DESCRIPTION
**Pull request by Raahul-Singh**
This is the python implementation of the ELO score.
I need to figure out how to test this. Running it on the entire dataset will take a long time, and we cannot verify it otherwise.
Probably need to create some dummy data.
Finally, this may have some minute bugs that may have escaped manual testing. These will be sorted after we figure out testing.

Fixes: #7

### CODE REVIEW

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: Why would be the starting point 1?

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: why would be the `std_dev` initialised with the `k_value`?

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: What was the motivation to keep the previous scores?

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: would it make more sense to use a boolean instead? And the argument being: `wins_0_over_1` or simpler: `wins_0`

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: if `score_for_image` is 0, that's then multiplied by a negative number. That's right, right?

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: You could use a [`deque`](https://docs.python.org/3/library/collections.html#collections.deque) of a `score_memory` length instead of having to check whether that lenght has be reached each time.

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]:
```suggestion
        new_std_dev_0 = min(np.std(state_dict_0['last_scores']), 1_000_000)  # prevents Infinity
```
To ease readability of this "magic number". And since it's a magic number, then probably better to assign it to a variable.

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: all right, I see now where the `last_scores` is needed.

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: This function repeats everything twice, how can we make it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)er?

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: I = 0... I thought you had a better self-esteem

File: pythia/seo/ELO.py
Comment (dpshelio) [inline pythia/seo/ELO.py]: :scream_cat: a hardcoded file name!!!

**Review by dpshelio — COMMENTED** [review]
Very good @Raahul-Singh I'm very happy with the quality shown in here! Well done. I've made some small comments (many are more of a conversation with myself :sweat_smile:), you don't need to agree with them all.

Finally, maybe we want to keep this more general, then we may break this into ELO only and ELO for sunspotter.

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: Ah, this was to avoid divide by zero errors that I was running into. I was calculating the std manually. Numpy handles it now.

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: It numerically is equal to k_value in the original code. 
I think it should be initialised as `score_change['max']`

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: We need these to calculate the standard deviation and update the k_value.

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: Yes, because the `image_expected_score` is always greater than 0

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: No, because we need the integers for calculating the `new_rating`
Numpy would convert the boolean to integers but it would just add unnecessary computations.

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: I did not know we could do this in python!

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: The only apt response to failing to notice this is, abject incessant groaning. And that has been me since I read your review comment.

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: ** abject incessant groaning intensifies **

File: pythia/seo/ELO.py
Comment (Raahul-Singh) [reply pythia/seo/ELO.py]: Done in the latest refactor!

**Comment by Raahul-Singh** [comment]
Rankings initially:
```
          image_id  score  k_value  count  std_dev last_scores
image_id                                                      
1                1   1400       32      0       32        1400
2                2   1400       32      0       32        1400
3                3   1400       32      0       32        1400
4                4   1400       32      0       32        1400
5                5   1400       32      0       32        1400
6                6   1400       32      0       32        1400
7                7   1400       32      0       32        1400
8                8   1400       32      0       32        1400
9                9   1400       32      0       32        1400
10              10   1400       32      0       32        1400

```
The matches are 
```
    Player 1 id      Player 2 id     does player 1 win? (1: Yes, 0: No)
           1           6                           1
           2           7                           0
           3           8                           0
           4           9                           1
           5          10                           1
```
Rankings after each player have had one match:
```    
      image_id   score  k_value  count  std_dev    last_scores
image_id                                                          
1                1  1416.0     16.0    1.0      8.0  1400.0,1416.0
2                2  1384.0     16.0    1.0      8.0  1400.0,1384.0
3                3  1384.0     16.0    1.0      8.0  1400.0,1384.0
4                4  1416.0     16.0    1.0      8.0  1400.0,1416.0
5                5  1416.0     16.0    1.0      8.0  1400.0,1416.0
6                6  1384.0     16.0    1.0      8.0  1400.0,1384.0
7                7  1416.0     16.0    1.0      8.0  1400.0,1416.0
8                8  1416.0     16.0    1.0      8.0  1400.0,1416.0
9                9  1384.0     16.0    1.0      8.0  1400.0,1384.0
10              10  1384.0     16.0    1.0      8.0  1400.0,1384.0

```

**Comment by Raahul-Singh** [comment]
@dpshelio if this looks good, I'll run it on the 14 years dataset and leave it to run overnight.

File: pythia/seo/elo.py
Comment (drsophiemurray) [inline pythia/seo/elo.py]: I suggest documenting what these numbers are for non-experts!

File: pythia/seo/elo.py
Comment (drsophiemurray) [inline pythia/seo/elo.py]: Change `expexted_0_score` to `expected_0_score`

**Review by drsophiemurray — COMMENTED** [review]
Nice tidy code @Raahul-Singh! Only a couple comments from me as I think @dpshelio covered most of it 👍

File: pythia/cleaning/elo.py
Comment (dpshelio) [inline pythia/cleaning/elo.py]: There's a style also on how python suggests to import things:
```
python std

external

internal
```
But... the good thing is that you don't need to do that manually! you can run [`isort`](https://github.com/timothycrosley/isort) and set the configuration in `setup.cfg`

File: pythia/cleaning/elo.py
Comment (dpshelio) [inline pythia/cleaning/elo.py]:
```suggestion
                                   f" are not present in the score board: {missing_columns}")
```

File: pythia/cleaning/elo.py
Comment (dpshelio) [inline pythia/cleaning/elo.py:75]: there are some varibles that refers as playerse and others as images. Maybe we should aim for a consistency across the whole class. You may want to leave this for the future after it's merged.

File: pythia/cleaning/elo.py
Comment (dpshelio) [inline pythia/cleaning/elo.py]: I think we can make all the above DRY-er - I'll make a PR

File: pythia/cleaning/elo.py
Comment (dpshelio) [inline pythia/cleaning/elo.py:190]: In the future we should start to move print messages using a logger.

File: pythia/cleaning/tests/test_elo.py
Comment (dpshelio) [inline pythia/cleaning/tests/test_elo.py]: If the file is only used in this test, I probably would save the file in this directory.

File: pythia/cleaning/tests/test_elo.py
Comment (dpshelio) [inline pythia/cleaning/tests/test_elo.py]: I wouldn't use the original dataset as for tests..

File: pythia/cleaning/tests/test_elo.py
Comment (dpshelio) [inline pythia/cleaning/tests/test_elo.py]: Normally you wouldn't use `==` for testing that many decimals. Rather use `pytest.approx` or `np.testing.assert_allclose`

File: pythia/cleaning/tests/test_elo.py
Comment (dpshelio) [inline pythia/cleaning/tests/test_elo.py:66]: Including raises! Well done!!

**Review by dpshelio — APPROVED** [review]
It seems good to me, so you can go ahead and merge it. You can create issues for the test data or names of variables to solve them later.

File: pythia/cleaning/elo.py
Comment (Raahul-Singh) [reply pythia/cleaning/elo.py:75]: #40

File: pythia/cleaning/elo.py
Comment (Raahul-Singh) [reply pythia/cleaning/elo.py:190]: #41

File: pythia/cleaning/tests/test_elo.py
Comment (Raahul-Singh) [reply pythia/cleaning/tests/test_elo.py]: #42

**Pull request merged**
